File: core/src/main/kotlin/fr/rob/core/network/v2/netty/client/NettyChannelHandler.kt
package fr.rob.core.network.v2.netty.client

import fr.rob.core.network.v2.ClientInterface
import fr.rob.core.network.v2.session.SessionSocketInterface
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.util.ReferenceCountUtil

abstract class NettyChannelHandler<T>(protected val client: ClientInterface<T>) : ChannelInboundHandlerAdapter() {

    abstract fun createPacketFromMessage(msg: Any): T
    abstract fun createSessionSocket(ctx: ChannelHandlerContext): SessionSocketInterface

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        try {
            client.onPacketReceived(createPacketFromMessage(msg))
        } catch (exception: Exception) {
            // TODO: use a logger instead
            exception.printStackTrace()

            if (ctx.channel().isOpen) {
                ctx.channel().close()
            }
        } finally {
            ReferenceCountUtil.release(msg)
        }
    }

    override fun channelActive(ctx: ChannelHandlerContext) {
        val session = client.createSession(createSessionSocket(ctx))

        client.onConnectionEstablished(session)
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        cause.printStackTrace()
        ctx.close()
    }
}

File: client/RobClient/src/network/NettyChannelHandler.cs
using System;
using DotNetty.Common.Utilities;
using DotNetty.Transport.Channels;
using Fr.Raven.Proto.Message.Gateway;

namespace RobClient.Network {
    class NettyChannelHandler : ChannelHandlerAdapter {
        private Client _client;
        
        public NettyChannelHandler(Client client) {
            _client = client;
        }
        
        public override void ChannelRead(IChannelHandlerContext context, object message) {
            _client.OnPacketReceived((Packet)message);

            ReferenceCountUtil.Release(message);
        }

        public override void ChannelActive(IChannelHandlerContext context) {
            _client.OnConnectionEstablished(context.Channel);
        }

        public override void ExceptionCaught(IChannelHandlerContext context, Exception exception) {
            Console.WriteLine(exception.ToString());
        }
    }
}
